### PR TITLE
Add int1 to int7 dtypes

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136301

Summary:
Similar to https://github.com/pytorch/pytorch/pull/117208, we want to add int1 to int7 for edge use cases
for weight quantization (https://www.internalfb.com/diff/D62464487)

Test Plan:
python test/test_quantization.py -k test_uint4_int4_dtype

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by ezyang** [comment]
Can you explain in more detail why the signed types are desirable?

**Comment by jerryzh168** [comment]
@ezyang it's used in edge low bit kernels to represent weight: https://www.internalfb.com/diff/D62464487

TBH, I think it we can extend torch.dtype out of core, it's probably better, but my impression is that we can't do that right now

**Comment by jerryzh168** [comment]
actually we can delay this PR I think, since we have been uint8 right now for uintx as well: https://github.com/pytorch/ao/blob/53b6b787bb035f1abc3c4804dd22c09ee41e7a0d/torchao/dtypes/uintx/uintx.py#L66, we just need to add some placeholder dtypes in torchao in this case.

**Comment by jerryzh168** [comment]
@ezyang you mean just for int1 or other intx dtypes as well? I agree int1 is probably not very useful. but cc @metascroy and also @kimishpatel to see if there are more context here for the need intx dtypes (especially int1).

**Comment by ezyang** [comment]
For the larger int types, it seems to me that the biggest benefit you get over unsigned is access to signed operations. But this also isn't obviously that helpful; for example, in two's complement, you can implemented signed addition... as regular unsigned addition.

Obviously, exposing signed integers as a user facing abstraction is a legitimate thing to do, but my understanding of AO's use case is they would wrapped up in a tensor subclass anyway...

File: c10/core/ScalarType.h [hunk 1]
```diff
@@ -467,14 +479,22 @@ inline bool isSignedType(ScalarType t) {
...
+    case ScalarType::Int7:
+      return true;
     case ScalarType::UInt1:
     case ScalarType::UInt2:
     case ScalarType::UInt3:
     case ScalarType::UInt4:
     case ScalarType::UInt5:
     case ScalarType::UInt6:
     case ScalarType::UInt7:
-      return true;
```
Comment (kimishpatel) [inline c10/core/ScalarType.h:477]: This seems like a bug before this pr?

File: test/quantization/core/test_utils.py [hunk 2]
```diff
@@ -192,30 +192,31 @@ def test_quantize_weight_clamping_per_channel(self):
         assert quantized_tensor.int_repr().max().item() == q8_max
         assert quantized_tensor.int_repr().min().item() == q8_min
 
-    def test_uint1_7_dtype(self):
+    def test_uint4_int4_dtype(self):
```
Comment (kimishpatel) [inline test/quantization/core/test_utils.py:195]: what happened to the original test case?

**Review by kimishpatel — COMMENTED** [review]
left some comments

**Comment by kimishpatel** [comment]
@ezyang so in general it is not clear why unsigned is treated different than signed? We have had this issue with uint16. So just from UX perspective it seems like an aberration that we have uintx types but not intx.

> but my understanding of AO's use case is they would wrapped up in a tensor subclass anyway

I think this is fair and I am not sure for the kernels we actually need intx/uintx types at all if kernels are tied with quantization API. Meaning if we did not use AO API, and relied purely on exported IR to derive if some tensors are uintx/intx, then I think having this natively available in PyTorch makes sense.

But in general @ezyang what would be the objection given uintx are already present.

File: c10/core/ScalarType.h (hunk 1 above)
Comment (jerryzh168) [reply c10/core/ScalarType.h:477]: yeah I think so

File: test/quantization/core/test_utils.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/core/test_utils.py:195]: naming is a bit off, original test is testing uint4, and I added int4 in the PR

**Comment by jerryzh168** [comment]
> I think this is fair and I am not sure for the kernels we actually need intx/uintx types at all if kernels are tied with quantization API. Meaning if we did not use AO API, and relied purely on exported IR to derive if some tensors are uintx/intx, then I think having this natively available in PyTorch makes sense.

yes all the operations are implemented with tensor subclass (currently just pack/unpack to/from uint8), we are using torch.uintx/torch.intx as user facing API so people don't need to maybe use (nbits + signed bool) to express them

**Review by ezyang — APPROVED** [review]

**Comment by jerryzh168** [comment]
@ezyang any ideas on the CI failure (https://github.com/pytorch/pytorch/actions/runs/11026420870/job/30727936288?pr=136301)? I can't repro locally

**Comment by ezyang** [comment]
looks like master problem

**Comment by jerryzh168** [comment]
landed in https://github.com/pytorch/pytorch/pull/137928

**Pull request closed without merging**